feat: Introduce Enabled Template to all overlays

Rafii2198 opened the pull request:
This is a continuation of #3373 which was closed due to targeting development branch and I had some issues with GitHub and rebasing, so I decided to just open a new PR. Description is copied from there.

In my wardrobe PR #3348, ShadowCat suggested a change to make all overlays a bit more standardized in terms on when they should render, as currently the only way we did that was simply early return in `render()` method. I decided to try to make such change, I don't know if this is the right way but in game it works really well, at least on surface level.

With this new change, all overlays as a baseline only render when on world and when player is not a spectator as well as give the ability for players to input a function in new Enabled Template field that was previously only present in Custom Bars and Text Overlays. The implementation of Enabled Template is a bit different, originally for those overlays it was the only condition on when it should be rendered, but as overlays have a default condition as baseline, it works in addition to it, so both need to be true in order for it to render, but there is also a new setting called Enabled Template Overwrite that when checked, it will cause enabled template to completely replace default condition.

Additionally, all overlays can have their special default condition, with the way I implemented it, all overlays add additional conditions on top of default one, except for BombBellOverlay which I made it completely overwrite the default one and render as long as there any bombs active, that means it should render in hub (which I couldn't test) making it easier to change worlds when not using the /world command.

The only downside I see is that we would need to make sure it doesn't crash in specific cases, like game bars were crashing when there was no bar active, and it tried to render which I had to fix, which was simple, but we have to keep an eye on that.

Lastly, no upfixers are needed, as the setting is still called `enabledTemplate`, their config keys haven't changed

Example of it in action, apologies for low quality, I have some weird technical issues

https://github.com/user-attachments/assets/75c3890e-76a8-4b06-bade-a15dcc2a9c15

Enabled Template of minimap was `at_cap(capped_mana)` meaning it should only render when my mana is at max which is the case, additionally as Enabled Template Overwrite is enabled, only Enabled Template condition is taken into consideration, meaning as long that function returns true it will render which is the case in character select
<img width="2560" height="1417" alt="image" src="https://github.com/user-attachments/assets/9205d18a-5b88-4964-bc34-9cb089782466" />

Rafii2198 committed:
feat: Introduce Enabled Template to all overlays

Rafii2198 committed:
remove unused i18n

Rafii2198 committed:
remove unused i18n from all files

Rafii2198 committed:
fix mention to bar instead of overlay

ShadowCat117 commented:
Oh whoops sorry didn't realise it would close your pr

Rafii2198 commented:
No problem, I had these issues with rebase anyway so I imagine I would close it myself anyway

magicus commented:
There is a lot of calls in the style of
```
 return super.defaultRenderCondition() && Models.Ability.manaBankBar.isActive();
 ```

This indicates to me that there is a design improvement hiding here. Something along the lines of changing the top-level `defaultRenderCondition` to include `&& isActive()` and having a default implementation of `isActive()` as `return true`. Then all the subclasses just need to override `isActive` with their specific limitations.

(Or something like that, perhaps with better names, but I hope you get the gist of it)

Rafii2198 commented:
But then overlay like Bombbell Overlay would also use the one from super, which in this case it doesn't, it has its own unique condition

magicus commented:
And then that class, and only that, need to override `defaultRenderCondition`. (Which probably should get a better name).

Rafii2198 commented:
Oh that actually makes sense, I kind of imagined making it private and the `isActive()` public. I think that makes sense, I will probably keep the name `defaultRenderCondition` and name the additional one `additionalRenderCondition` that way it should be pretty clear when an overlay "adds" to the condition and when it fully overwrites it. But if you have better suggestion for the name, I can change them to that too, I am not that good at naming things myself lol

Rafii2198 committed:
Merge remote-tracking branch 'refs/remotes/master/main' into enabled_template_for_all

Rafii2198 committed:
feat: introduce additionalRenderCondition()

Rafii2198 committed:
change to character wardrobe

Rafii2198 commented:
I don't know what happened, but since I made that PR, stuff like Character Wardrobe does not put you into spectator anymore somehow, at least it doesn't detect that which is weird, meaning I had to change that condition to inCharacterWardrobe specifically

Rafii2198 committed:
check for riding Display entity

Rafii2198 commented:
Actually I tried following Coehlrich's advice for checking riding Item Display entity, and it was much easier to implement than I imagined and it works! I only tried character wardrobe and opening a crate, but I would imagine it will work in all other similar cases.

magicus commented:
Why is bomb bell overriding the default render condition? Do you want to see it even when you are not on a world, when you are riding an entity or when the player is not setup? Could it not use the "additional" conditions as well? If so, we would only need a single method to override.

I'm not really happy with the names either. But first we need to figure out if we really need two methods.

Rafii2198 commented:
I would imagine it would be more beneficial to see active bombs regardless of your state so you won't miss them. Perhaps also Server Uptime overlay could make use of that to render when in hub so you know which world is newest. This also allows us to make specific overlays that appear in these scenarios which usually hide them, for example something specific to Character Wardrobe or something regarding Lootrun rewards that would be seen during the opening animation, it just gives us a possibility when needed. Also, just as I was writing it I realized our custom player list would need this as you can't see it in these cases.

magicus commented:
Even in the character selection screen?

ShadowCat117 commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +76,35 @@ protected Overlay(
 
+    /**
+     * <li>The default condition that determines whether an overlay is going to be rendered.</li>
+     * <li>Overwrite this to completly change the condition, or overwite additionalRenderCondition() instead to only add to the default condition.</li>
```
```suggestion
     * <li>Overwrite this to completely change the condition, or overwrite additionalRenderCondition() instead to only add to the default condition.</li>
```

ShadowCat117 commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +76,35 @@ protected Overlay(
+    /**
+     * <li>The default condition that determines whether an overlay is going to be rendered.</li>
+     * <li>Overwrite this to completly change the condition, or overwite additionalRenderCondition() instead to only add to the default condition.</li>
+     * <li>By default overlays only render when on world and not riding Display enity</li>
```
```suggestion
     * <li>By default overlays only render when on world and not riding a Display entity</li>
```

ShadowCat117 reviewed (approved):
I think this looks good overall, just minor spelling corrections.

And sorry for the delay, been focusing on the 2.1.3 fixes

Rafii2198 committed:
Merge branch 'main' into enabled_template_for_all

Rafii2198 committed:
Change ServerUptimeOverlay condition

Rafii2198 commented:
> I think this looks good overall, just minor spelling corrections.
> 
> And sorry for the delay, been focusing on the 2.1.3 fixes

I haven't had enough sleep and it shows, lol.
Also, it's fine, like I said on discord, this can wait as getting the mod working is more important than such quality of life changes.

> Even in the character selection screen?

Yes, I think it would be beneficial for those who can see it, after all this only works for people who have Champion rank, I would assume they want to see it and if they don't then they can change it to whatever they want with functions.

Rafii2198 committed:
add condition CustomPlayerListOverlay

Rafii2198 committed:
Merge remote-tracking branch 'origin/enabled_template_for_all' into enabled_template_for_all

Rafii2198 committed:
fix Javadoc

magicus commented:
I am still not really happy with the names  defaultRenderCondition() and additionalRenderCondition(). First, I think it should be like a verb, shouldRender or something like that. Then I don't like "default" and "additional", it does not really say anything useful. 

I am not able to come up with better names on the spot, unfortunately. Maybe @kristofbolyai can chime in here?

Rafii2198 commented:
How about `isRenderedDefault()` and `isRenderedAdditional()`? The first one was the name that TextOverlay used.

This was also one of the reasons I did it as just one method, splitting it into two needs some unclear naming, and it can be confused, when it is only one which calls the super I believe it's easier to understand as it is read as "default condition and this special condition", with the split there is a possibility that you misread or confuse which one is overwritten or which one you want to overwrite which can lead to unexpected results. At least in my opinion, of course.

Rafii2198 committed:
Merge remote-tracking branch 'master/main' into enabled_template_for_all

# Conflicts:
#	common/src/main/java/com/wynntils/overlays/gamebars/FocusedMobHealthBarOverlay.java

kristofbolyai commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +76,35 @@ protected Overlay(
         this.verticalAlignmentOverride.store(verticalAlignmentOverride);
     }
 
+    /**
```
These javadocs seems to be a bit non-standard.  I don't think you need the `<li>`s, do you?

Rafii2198 replied:
I personally felt that having these 2 methods could be confusing, so I wanted to make the doc as clear as possible

kristofbolyai replied:
Adding a list or not does not change anything for understandability. Javadocs already respect newlines iirc (you may have to put 1 empty line in between content lines)

kristofbolyai commented on common/src/main/java/com/wynntils/core/consumers/overlays/OverlayManager.java:
```diff
@@ -205,7 +205,10 @@ private void recreateGroupOverlaysWithIds(OverlayGroupHolder holder, List<Intege
-        enabledOverlays.forEach(Overlay::tick);
+        enabledOverlays.forEach(overlay -> {
+            overlay.tick();
+            overlay.updateCache();
```
This is reallyyyy not supposed to be updated on every tick. It just needs to be done in `updateConfigOption`. The only reason your game could still run with this is because function manager already does caching on function evaluation.

Rafii2198 replied:
But then wouldn't the function be evaluated only when configs are updated? Like for example if I set the condition to be when I am at max mana, wouldn't it only be evaluated when I save configs?

kristofbolyai replied:
Yes, my bad.

kristofbolyai commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +76,35 @@ protected Overlay(
+     */
+    protected boolean additionalRenderCondition() {
+        return true;
+    }
```
Why two methods? If I want additional conditions, I override and call super and-ed with my other conditions.

Rafii2198 replied:
Basically the whole conversation I had here with Magicus, originally that was just one method

kristofbolyai replied:
Okay, I get that. This just demonstrates that I also don't like the current name and my brain automatically wanted to just get rid of this all together. I overall feel like that we are so overwhelming the simple enabled / disabled concept that it is getting hard for even us to understand. Can we spend a bit of time thinking through what we really need?

Rafii2198 replied:
Having just one method is sufficient, but it will cause many calls to super as for the most part overlays want to extend that condition with their own specifics, but not always like with custom player list which only needs to be shown when tab list would be displayed. Functionally only this is needed. The reason there are 2 different once is that Magicus believes that the calls to super are common and thus could be improved by splitting it into 2 methods that would be both checked when evaluating.

If it were up to me, I would revert to just 1 single method, and call it `isRenderedDefault()` like our bars call it. This would go back to square one where there are common calls to super, but I believe it's better than this split. Having them split also doesn't allow for things like having OR condition with the default one, like when you wanted to have it rendered as normal or when something else is happening, currently with this split both of them need to be true.
Unless these frequent calls to super cause some technical issues that I am not aware of and misunderstood what Magicus truly wanted.

kristofbolyai commented on common/src/main/java/com/wynntils/core/consumers/overlays/TextOverlay.java:
```diff
@@ -123,27 +119,4 @@ private float getTextScale() {
-     * @return whether the overlay is rendered with the default (empty) template
-     */
-    public boolean isRenderedDefault() {
-        return Models.WorldState.onWorld() && !Models.WorldState.inCharacterWardrobe();
```
For a bunch or overlays you removed the `!inCharacterWardrobe` check. Why?

Rafii2198 replied:
I made a better check, it basically checks if you are riding a display entity which happens during these moments which includes character wardrobe

kristofbolyai commented on common/src/main/java/com/wynntils/overlays/CustomPlayerListOverlay.java:
```diff
@@ -76,10 +76,14 @@ public void onRender(RenderEvent.Pre event) {
 
+    @Override
+    protected boolean defaultRenderCondition() {
+        return !(!McUtils.options().keyPlayerList.isDown() && animationPercentage.finishedClosingAnimation());
```
```suggestion
        return McUtils.options().keyPlayerList.isDown() || !animationPercentage.finishedClosingAnimation();
```

magicus replied:
That is even better expressed as two separate return statements.

magicus replied:
Actually, I'm stupid. :) I mis-read it as `if (McUtils.options().keyPlayerList.isDown() || !animationPercentage.finishedClosingAnimation()) return;`. Had that been the case, two separate return statements had been better. 

For this, I agree with bolyai's de-morgan-ization.

magicus commented:
Try this instead then: Add another method, `hideWhenNoGui` which by default returns true. Override it to return false at the place where you wanted to override defaultRender. Inline the logic of defaultRender into place, and couple it with a `&& !hideWhenNoGui()`.

Rename "defaultAdditionalRender" to "isVisible".

Rafii2198 commented:
But then wouldn't it be the original issue with calls to super when overlays want to add onto the default?

magicus commented:
Then they will override `isVisible`. (just a new name for additionalRender)

Rafii2198 committed:
add hideWhenNoGui and rename methods

Rafii2198 committed:
Merge branch 'main' into enabled_template_for_all

magicus commented:
The new `hideWhenNoGui` and `isVisible` works very well on the subclasses. It is now quite clear what they are trying to express, and in several cases much better than what was there before. Nice!

However, I'd still like to see some more work on the base class. I'll comment inline.

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -9,19 +9,32 @@
+    @Persisted(i18nKey = "overlay.wynntils.overlay.enabledTemplate")
+    protected final Config<String> enabledTemplate = new Config<>("");
+
+    private ErrorOr<Boolean> enabledTemplateCache = null;
```
Having a cache complicates things. Is this really necessary? We did not have a cache before, and we are basically doing the same kind of checks for the same kind of cost.

Did you measure anything to conclude we needed a cache, or was it just like "well it felt like it was needed" profiling?

Edit: Aha, now I see. This is for parsing of templates. Then it makes sense to have a cache, yes.

magicus replied:
However, please move this line to after all `@Persisted`.

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +76,37 @@ protected Overlay(
+
+    /**
+     * Whether the overlay should be hidden when Wynncraft hides gui.
+     * @return true
```
```suggestion
```

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +76,37 @@ protected Overlay(
+
+    /**
+     * Whether the overlay should be rendered.
+     * @return true
```
```suggestion
```

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -9,19 +9,32 @@
+    private ErrorOr<Boolean> enabledTemplateCache = null;
+
+    @Persisted(i18nKey = "overlay.wynntils.overlay.enabledTemplateOverwrite")
+    protected final Config<Boolean> enabledTemplateOverwrite = new Config<>(false);
```
I think the `enabledTemplateOverwrite` option has multiple issues. Is it not just enough to assume that if a user takes control of when an overlay is enabled, they will have the full power to dictate it?

The  descriptions of these two new options is also not very helpful:
```
  "overlay.wynntils.overlay.enabledTemplate.description": "A template returning a boolean value, which determines whether this overlay should be rendered or not. This template should not have curly brackets. This works in addition to default rendering condition of this overlay",
  "overlay.wynntils.overlay.enabledTemplateOverwrite.description": "Whether the Enabled Template should completly overwrite default rendering condition of this overlay rather than adding to it",
```
but what you have *actually* coded is that the enabledTemplate will be ANDed to the normal logic. So you can only **turn off** the overlay more, but you can never **turn on** it more. It is not clear that this is what means by " in addition to default rendering condition".

Or, if you think that this is what is really needed, a way to hide an overlay when it would otherwise be shown, but not a way to show the overlay when it would otherwise be hidden, then we can also drop `enabledTemplateOverwrite` and just keep the default behavior, and clarify the documentation that this can only be used to turn off the overlay.

Rafii2198 replied:
I don't know how else I can describe them, in my opinion it's fairly understandable, but I could be wrong. The intent is to make enabled template be an additional condition that needs to be met on top of the default condition, but when to overwrite is enabled then the default condition is completely disregarded and enabled template is the only thing controlling when it should be visible.

magicus replied:
As a general rule, if a variable or function is hard to describe, it is most often capturing the wrong aspect. It usually helps to take a step back and see if there is a clearer concept that could be brought forward that solves the same problem.

In this case, you could do like the "hideWhenNoGui". If you invert the value of the template, it becomes much clearer: if the value were named e.g. `hideOverlayTemplate` then it is clear that if it returns true, the overlay will be hidden -- but it does not imply that it is necessarily shown when it is false.

And instead of a big boolean making a single variable play two different roles, you could have an `enabledTemplate`, which -- if set -- completely overrides the normal Wynntils logic.

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +76,37 @@ protected Overlay(
         this.verticalAlignmentOverride.store(verticalAlignmentOverride);
     }
 
+    private boolean isRenderedDefault() {
```
I don't like this function. I think it should be inlined into isRendered. You will need to rewrite it a bit to do it without having to duplicate code or run stuff unnecessary. This might happen anyway if you decide to drop enabledTemplateOverwrite.

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/OverlayManager.java:
```diff
@@ -205,7 +205,10 @@ private void recreateGroupOverlaysWithIds(OverlayGroupHolder holder, List<Intege
-        enabledOverlays.forEach(Overlay::tick);
+        enabledOverlays.forEach(overlay -> {
+            overlay.tick();
+            overlay.updateCache();
```
Should we really update the cache on every tick? That seems kind of excessive. But maybe that is what we do with normal template rendering?

magicus replied:
And if we should, then maybe we should move the call to updateCache inside the tick method in Overlay, since it is then a necessary part of Overlays to function correctly. That also means we must make sure all overrides if that tick function start by calling `super.tick()`. (And since that is easy to forget, it is an argument against my suggestion and for keeping it like it is.)

Rafii2198 replied:
This was my line of thinking for making it a separate method, updating cache cannot be split and it just so happens on tick, it could be done each frame, that would remove the need for cache, but I don't know if it will cause performance issues.

magicus replied:
Fine. But please rename the method to `updateEnabledCache` to clarify it's purpose.

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -78,6 +122,18 @@ public void renderPreview(
+
+        String formattedTemplate =
+                StyledText.join("", Managers.Function.doFormatLines(template)).getString();
+        this.enabledTemplateCache = Managers.Function.tryGetRawValueOfType(formattedTemplate, Boolean.class);
```
I assume tryGetRawValueOfType returns an error which is checked by enabledTemplateCache.hasError, if the user makes a mistake. But how will we propagate this error to the user?

I have not checked but I assume that if the normal template value returns an error, we will write a red "SYNTAX ERROR IN TEMPLATE" on the screen, or something like that. Here, an error will just go unnoticed, except that it will just not work.

I think that if the enabled template has an error, we should not print the normal text but instead show "error in template". Which does not work for other Overlays, I realize. Maybe just render a text "Error in enabled template" on top of the overlay.

Rafii2198 replied:
I believe this would need to be done in its own PR as the enabled template itself is not a new thing, I just brought it from specific overlays to all overlays as a core. I believe we want to add a special screen for writing functions, that will show errors and help.

magicus replied:
But when doing that you also promote it to a core Overlay functionality, which will warrant more careful and thorough work. But fair enough, we can handle that in a future PR. I would like you to note in a comment that this will hide any user errors.

Rafii2198 committed:
merge isRenderedDefault and isRendered methods

Rafii2198 committed:
remove @return from Javadocs

Rafii2198 committed:
move enabledTemplateCache after all @Perstited

Rafii2198 committed:
remove enabledTemplateOverwrite

Rafii2198 committed:
Implement error rendering

Rafii2198 committed:
Merge remote-tracking branch 'refs/remotes/master/main' into enabled_template_for_all

Rafii2198 commented:
Apologies for the delay. I implemented your suggested fixes and hopefully didn't miss anything.
I followed Magicus advices and made 2 changes, first I removed the overwrite setting, instead Enabled Template just replaced default rendering from the get-go. Second, I got convinced and decided to try to actually implement error rendering rather than putting it for later with a bigger change, this is how it looks:
<img width="2560" height="1417" alt="image" src="https://github.com/user-attachments/assets/fa0a2c48-2e32-47aa-8016-60fd0513afd3" />
When the user provides an Enabled Template, and it has an error, instead of normal render it will render this error message

ShadowCat117 commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +76,30 @@ protected Overlay(
+        if (enabledTemplateCache == null || enabledTemplateCache.hasError())
+            return ((Models.WorldState.onWorld()
+                                    && (McUtils.player() != null
+                                            && !(McUtils.player().getVehicle() instanceof Display)))
```
If you're up for it, I think LocalPlayer, or one of the super classes of it, has a method for riding and exiting vehicles. Perhaps we should mixin to those calls and set the vehicle type in a model somewhere.

Then we can expose this as a function for enabled templates to still consider this. (And also a function for riding a horse has been requested occasionally)

Rafii2198 replied:
I'll take a look at it. Just to get it right, add a field like `Entity vehicle` to CharacterModel, then set it to the entity the player is riding or null when none is being ridden and check if that entity is instanceof Display here, right?

ShadowCat117 replied:
I'd do the instanceof check in the event handling and then just make an enum of vehicletype as we're probably only interested in Display and Horse

Rafii2198 replied:
Hmm it happens in Entity class, so I would need to check if LocalPlayer is calling it, but I am not sure how to check for that in a mixin

ShadowCat117 replied:
LocalPlayer does override it

```
@Override
public boolean startRiding(Entity vehicle, boolean force) {
    if (!super.startRiding(vehicle, force)) {
        return false;
    } else {
        if (vehicle instanceof AbstractMinecart) {
            this.minecraft.getSoundManager().play(new RidingMinecartSoundInstance(this, (AbstractMinecart)vehicle, true));
            this.minecraft.getSoundManager().play(new RidingMinecartSoundInstance(this, (AbstractMinecart)vehicle, false));
        }

        return true;
    }
}

@Override
public void removeVehicle() {
    super.removeVehicle();
    this.handsBusy = false;
}
```

Rafii2198 replied:
Hopefully I did it right, it works in game

ShadowCat117 commented on common/src/main/java/com/wynntils/core/consumers/overlays/OverlayManager.java:
```diff
@@ -272,9 +275,15 @@ private void renderOverlays(RenderEvent event, RenderState renderState) {
-                    overlay.render(event.getGuiGraphics(), BUFFER_SOURCE, event.getDeltaTracker(), event.getWindow());
+                    if (overlay.enabledTemplateCache != null && overlay.enabledTemplateCache.hasError()) {
+                        overlay.renderErrorMessage(event.getGuiGraphics(), BUFFER_SOURCE);
+
```
```suggestion
```

Rafii2198 committed:
remove empty line

Rafii2198 committed:
implement new event

Rafii2198 committed:
update translation in error message

Rafii2198 committed:
fix isVisible for CustomPlayerListOverlay

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +75,28 @@ protected Overlay(
+            return ((Models.WorldState.onWorld() && Models.Character.getVehicle() != VehicleType.DISPLAY)
+                            || !hideWhenNoGui())
+                    && isVisible();
+        return enabledTemplateCache.getValue();
```
Please add {} after if. Braceless if statements are only allowed for returns (or break/continue) and only if it fits on a single line.

edit: Actually, I think the entire function can be rewritten...

I find this code very hard to parse.  If you split it up, I think it will become much easier to read. Something like this:

```suggestion
        if (enabledTemplateCache != null && !enabledTemplateCache.hasError()) {
            return enabledTemplateCache.getValue();
        }
        
        if (!isVisible()) return false;
        if (hideWhenNoGui()) {
            return Models.WorldState.onWorld() && Models.Character.getVehicle() != VehicleType.DISPLAY;
        }
        return true;
```

Rafii2198 replied:
I might be misreading, but separating isVisible from hideWhenNoGui might lead to cases where there is no gui, but the overlay is still visible

Rafii2198 replied:
Ah yeah that's my bad, I was misreading, sorry

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -76,8 +110,43 @@ public void renderPreview(
         this.render(guiGraphics, bufferSource, deltaTracker, window);
     }
 
+    public void renderEnabledTemplateErrorMessage(GuiGraphics guiGraphics, MultiBufferSource bufferSource) {
```
This should move to OverlayManager and be made private.

magicus replied:
... unless you intend to overwrite it in some Overlay subclasses?

magicus replied:
But in that case at least make it package-private (that is, no visibility qualifier at all).

magicus replied:
Hm. Or maybe it need to be protected in that case.

kristofbolyai replied:
Yes, make this protected.

kristofbolyai replied:
You can also make `tick` protected.

magicus replied:
Actually, I think it should be made private with the new `renderOrErrorMessage`. If we ever want to override this we can loosen it up.

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/OverlayManager.java:
```diff
@@ -272,9 +275,14 @@ private void renderOverlays(RenderEvent event, RenderState renderState) {
+                } else if (shouldRender && overlay.isRendered()) {
                     long startTime = System.currentTimeMillis();
-                    overlay.render(event.getGuiGraphics(), BUFFER_SOURCE, event.getDeltaTracker(), event.getWindow());
+                    if (overlay.enabledTemplateCache != null && overlay.enabledTemplateCache.hasError()) {
```
Actually, now I finally figured out how this should be resolved. The problem is that you expose the enabledTemplateCache to the OverlayManager. This should be internal to the Overlay. So the decision whether to render an error message or the real overlay should be made by the Overlay itself. Since overlays override the `render` method, you will need to introduce a new top-level method in Overlay, e.g. `doRender`, that does this check and selects between rendering an error or delegating to the Overlay's `render` method.

magicus commented:
I just want to add that this is starting to look really, really good! Thank you for hanging in there @Rafii2198 and help making this well-designed!

Rafii2198 committed:
move rendering check to overlay itself

ShadowCat117 commented on common/src/main/java/com/wynntils/mc/mixin/LocalPlayerMixin.java:
```diff
@@ -35,4 +37,16 @@ private void playSoundPre(SoundEvent sound, float volume, float pitch, CallbackI
         }
     }
+
+    @Inject(method = "startRiding", at = @At("HEAD"))
```
Be sure to use the qualified name for both of these (if you have the mc dev plugin, right click the method in LocalPlayer, copy/paste special, mixin target reference. That will include the class before it like Lnet/minecraft/.../LocalPlayer but you can remove that)

Rafii2198 replied:
When I first did it for Entity mixin it actually autocompleted it like that, but then when I remade this in LocalPlayerMixin, it didn't it just left plane name with nothing else

ShadowCat117 commented on common/src/main/java/com/wynntils/mc/mixin/LocalPlayerMixin.java:
```diff
@@ -35,4 +37,16 @@ private void playSoundPre(SoundEvent sound, float volume, float pitch, CallbackI
+        MixinHelper.post(event);
+    }
+
+    @Inject(method = "removeVehicle", at = @At("Head"))
```
```suggestion
    @Inject(method = "removeVehicle", at = @At("HEAD"))
```

Rafii2198 committed:
refactor isRendered method

Rafii2198 committed:
fix LocalPlayerMixin

Rafii2198 commented:
> I just want to add that this is starting to look really, really good! Thank you for hanging in there @Rafii2198 and help making this well-designed!

Thank you so much, it means a lot to me! This PR is one of the bigger ones from me and I have made so many new things and seeing it getting closer to working makes me really happy

magicus reviewed (approved):
I think this is okay now but I've been staring at it for so long so I might miss something. I'd appreciate another review as well.

kristofbolyai commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +75,37 @@ protected Overlay(
+        return true;
+    }
+
+    protected boolean isRendered() {
```
I think you should make this final. And check and fix `TokenBarsOverlay` overriding this method instead of the new options.

kristofbolyai commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +75,37 @@ protected Overlay(
+        if (hideWhenNoGui()) {
+            return Models.WorldState.onWorld() && Models.Character.getVehicle() != VehicleType.DISPLAY;
+        }
+        return true;
```
```suggestion
        // Otherwise render it according to defaults
        if (!isVisible() || !hideWhenNoGui()) return false;
        
        return Models.WorldState.onWorld() && Models.Character.getVehicle() != VehicleType.DISPLAY;
```

magicus replied:
The code above suggested by me. I still think it is easier to read than your suggestion. It reads like how you would pronounce it "if it is not visible, don't show it. if we should hide it when the gui is hidden, don't show it if we don't have a gui, otherwise show it".

But maybe it's just me.

magicus replied:
A bit of clarification could be to create `boolean hasGui = Models.WorldState.onWorld() && Models.Character.getVehicle() != VehicleType.DISPLAY`, and use that in combination with `hideWhenNoGui`.

Rafii2198 replied:
I don't know if I did what you wanted, I could have misunderstood this

magicus replied:
No misunderstanding, it looked good!

kristofbolyai commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -63,6 +75,37 @@ protected Overlay(
+        // When user provides Enabled Template but there is an error, render it to show the error
+        if (enabledTemplateCache != null && enabledTemplateCache.hasError()) return true;
+
+        // But if there are no errors, render acording to the template
```
```suggestion
        // But if there are no errors, render according to the template
```

kristofbolyai commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -76,8 +119,52 @@ public void renderPreview(
 
+    protected void updateCache() {
+        String template = enabledTemplate.get();
+        if (template.trim().isEmpty()) {
```
```suggestion
        if (template.isBlank()) {
```

kristofbolyai reviewed (commented):
With these mostly code quality based reviews I am ready to approve this once fixed. This is looking really good now.

magicus commented on common/src/main/java/com/wynntils/core/consumers/overlays/Overlay.java:
```diff
@@ -76,8 +119,52 @@ public void renderPreview(
         this.render(guiGraphics, bufferSource, deltaTracker, window);
     }
 
+    public void renderOrErrorMessage(
```
This can be package private (no qualifier), or at least protected.

Rafii2198 committed:
apply suggested changes

Rafii2198 committed:
rename updateCache() to updateEnabledCache()

magicus reviewed (approved):
I think this is finally good to be committed! 🥳

kristofbolyai committed:
Merge branch 'main' into enabled_template_for_all

kristofbolyai committed:
Merge branch 'main' into enabled_template_for_all
